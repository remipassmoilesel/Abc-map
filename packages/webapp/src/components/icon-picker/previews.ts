/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { IconProcessor } from '../../core/geo/styles/IconProcessor';
import { PointIcon } from '../../assets/point-icons/PointIcon';
import { IconCategory } from '../../assets/point-icons/IconCategory';
import { getAllIcons } from '../../assets/point-icons/helpers';

export interface IconPreview {
  icon: PointIcon;
  /**
   * Base 64 string of icon
   */
  preview: string;
}

let previews: Map<IconCategory, IconPreview[]> | undefined;

export function getPreviews() {
  if (!previews) {
    // We prepare previews
    const iconList = getAllIcons().map((i) => ({ icon: i, preview: IconProcessor.get().prepareCached(i, 50, '#0077b6') }));

    // We sort them by category
    previews = new Map<IconCategory, IconPreview[]>();
    for (const preview of iconList) {
      const category = previews.get(preview.icon.category);
      if (!category) {
        previews.set(preview.icon.category, [preview]);
      } else {
        category.push(preview);
      }
    }
  }

  return previews;
}
