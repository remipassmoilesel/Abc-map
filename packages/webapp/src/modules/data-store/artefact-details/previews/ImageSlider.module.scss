/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

@import 'src/styles/variables';

.slider {
  position: relative;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;

  cursor: pointer;

  background-color: $gray-100;
  border-radius: $border-radius;

  img {
    max-width: 100%;
    max-height: 100%;
  }
}

.loading {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;

  width: 100%;
  height: 100%;
}

.button {
  position: absolute;

  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;

  background: none;
  box-shadow: $box-shadow;
  border-radius: 50%;
  border: none;
  padding: 0;
  margin: 0;

  &.left {
    top: 50%;
    transform: translate(0, -50%);
    left: 1rem;
  }

  &.right {
    top: 50%;
    transform: translate(0, -50%);
    right: 1rem;
  }

  &:disabled {
    :global(.abc-icon) {
      display: none;
    }
  }
}

.attributions {
  position: absolute;
  bottom: 0.1rem;
  left: 0.1rem;
  padding: 0.4rem;
  font-size: 0.9rem;

  background-color: #ffffffaa;
  border-radius: $border-radius;
}
