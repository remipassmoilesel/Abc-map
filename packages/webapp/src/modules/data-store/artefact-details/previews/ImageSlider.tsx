/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import Cls from './ImageSlider.module.scss';
import { MouseEvent, useEffect } from 'react';
import { FaIcon } from '../../../../components/icon/FaIcon';
import { IconDefs } from '../../../../components/icon/IconDefs';
import { withTranslation } from 'react-i18next';
import clsx from 'clsx';
import { useCallback, useState } from 'react';
import { prefixedTranslation } from '../../../../i18n/i18n';
import { useServices } from '../../../../core/useServices';
import { Logger } from '@abc-map/shared';
import { BlueLoader } from '../../../../components/blue-loader/BlueLoader';
import { resolveInAtLeast } from '../../../../core/utils/resolveInAtLeast';

const logger = Logger.get('ImageSlider.tsx');

const t = prefixedTranslation('DataStoreModule:');

interface Props {
  urls: string[];
  attributions: string;
  onClick: () => void;
  className?: string;
  buttonSize?: string;
}

function ImageSlider(props: Props) {
  const { dataStore, toasts } = useServices();
  const { urls, attributions, onClick, className, buttonSize } = props;
  const [activeIndex, setActiveIndex] = useState<number>(0);
  const [activePreview, setActivePreview] = useState<string | undefined>();
  const [loading, setLoading] = useState(false);
  const hasPrevious = activeIndex > 0;
  const hasNext = activeIndex < urls.length - 1;

  // Download active preview
  useEffect(() => {
    if (!urls.length) {
      return;
    }

    let objectUrl: string | undefined;
    setLoading(true);
    resolveInAtLeast(dataStore.downloadFile(urls[activeIndex]), 200)
      .then((res) => {
        objectUrl = URL.createObjectURL(res.content);
        setActivePreview(objectUrl);
      })
      .catch((err) => {
        toasts.genericError(err);
        logger.error('Cannot download preview: ', err);
      })
      .finally(() => setLoading(false));

    return () => {
      if (objectUrl) {
        URL.revokeObjectURL(objectUrl);
      }
    };
  }, [activeIndex, dataStore, toasts, urls]);

  // Change current image, in specified direction. +1 for next, -1 for previous.
  const changePreview = useCallback(
    (direction: number) => {
      if (!urls.length) {
        return;
      }

      let index = activeIndex + direction;
      if (index < 0) {
        index = 0;
      }
      if (index >= urls.length) {
        index = urls.length - 1;
      }

      setActiveIndex(index);
    },
    [activeIndex, urls]
  );

  const handlePrevious = useCallback(
    (ev: MouseEvent) => {
      ev.stopPropagation();
      changePreview(-1);
    },
    [changePreview]
  );

  const handleNext = useCallback(
    (ev: MouseEvent) => {
      ev.stopPropagation();
      changePreview(1);
    },
    [changePreview]
  );

  return (
    <div className={clsx(Cls.slider, className)}>
      {/* Loading animation */}
      {loading && (
        <div className={Cls.loading}>
          <BlueLoader />
        </div>
      )}

      {/* Preview */}
      {!loading && activePreview && <img src={activePreview} alt={t('Preview_of_artifact')} onClick={onClick} />}

      {/* Control buttons */}
      {urls.length > 1 && (
        <>
          <button className={clsx(Cls.button, Cls.left)} onClick={handlePrevious} disabled={!hasPrevious}>
            <FaIcon icon={IconDefs.faArrowCircleLeft} size={buttonSize || '2rem'} />
          </button>
          <button className={clsx(Cls.button, Cls.right)} onClick={handleNext} disabled={!hasNext}>
            <FaIcon icon={IconDefs.faArrowCircleRight} size={buttonSize || '2rem'} />
          </button>
        </>
      )}

      <div className={Cls.attributions} dangerouslySetInnerHTML={{ __html: attributions }} />
    </div>
  );
}

export default withTranslation()(ImageSlider);
