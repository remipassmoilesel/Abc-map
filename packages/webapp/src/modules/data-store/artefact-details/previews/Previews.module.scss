/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

@import 'src/styles/variables';

$height: 30vh;

.previews {
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  justify-content: flex-start;

  // Some elements are absolutely positioned
  position: relative;

  height: $height;
  min-height: $height;
  max-height: $height;
}

.smallPreview {
  height: $height;
  max-height: $height;
  width: 100%;
  max-width: 100%;
}

.noPreview {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;

  min-width: 25rem;
  width: 25vw;
  height: 100%;

  border: solid 0.1rem $gray-400;
  border-radius: $border-radius;
}

.fullscreenModal {
  position: fixed;
  top: 0;
  right: 0;
  width: 100%;
  height: 100%;

  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;

  h3 {
    color: white;
    margin-bottom: 2rem;
  }

  background: rgba(0, 0, 0, 0.5);

  z-index: $zIndex-datastore-fullscreenPreview;
}

.largePreview {
  height: 75vh;
  max-height: 75vh;
  width: 80vw;
  max-width: 80vw;
}
