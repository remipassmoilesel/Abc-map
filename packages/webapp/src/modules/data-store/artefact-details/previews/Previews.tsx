/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import Cls from './Previews.module.scss';
import { AbcArtefact, getTextByLang, Logger } from '@abc-map/shared';
import { withTranslation } from 'react-i18next';
import { FaIcon } from '../../../../components/icon/FaIcon';
import { IconDefs } from '../../../../components/icon/IconDefs';
import { useCallback, useState } from 'react';
import clsx from 'clsx';
import { getLang, prefixedTranslation } from '../../../../i18n/i18n';
import { attributionsVariableExpansion } from '../../../../core/utils/variableExpansion';
import ImageSlider from './ImageSlider';

const t = prefixedTranslation('DataStoreModule:');

const logger = Logger.get('Previews');

interface Props {
  artefact: AbcArtefact;
  className?: string;
}

function Previews(props: Props) {
  const { artefact, className } = props;
  const [fullscreenPreview, showFullScreenPreview] = useState(false);
  const artefactName = getTextByLang(artefact.name, getLang());
  const attributions = attributionsVariableExpansion(artefact.attributions).join('\n');
  const previewUrls = artefact.previews || [];

  const toggleFullScreenPreview = useCallback(() => showFullScreenPreview(!fullscreenPreview), [fullscreenPreview]);

  return (
    <div className={clsx(Cls.previews, className)}>
      {/* Small preview */}
      {!!previewUrls.length && <ImageSlider urls={previewUrls} attributions={attributions} onClick={toggleFullScreenPreview} className={Cls.smallPreview} />}

      {/* No preview available*/}
      {!previewUrls.length && (
        <div className={Cls.noPreview}>
          <FaIcon icon={IconDefs.faImages} size={'4rem'} className={'mb-3'} />
          <h5>{t('No_preview_available')}</h5>
        </div>
      )}

      {/* Fullscreen preview */}
      {fullscreenPreview && (
        <div className={Cls.fullscreenModal} onClick={toggleFullScreenPreview} key={artefact.id}>
          <h3>{artefactName}</h3>
          <ImageSlider urls={previewUrls} attributions={attributions} onClick={toggleFullScreenPreview} buttonSize={'3rem'} className={Cls.largePreview} />
        </div>
      )}
    </div>
  );
}

export default withTranslation()(Previews);
