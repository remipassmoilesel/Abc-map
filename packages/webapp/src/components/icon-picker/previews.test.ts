/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { getPreviews } from './previews';

describe('previews', function () {
  it('should work', () => {
    const previews = getPreviews();

    const keys = Array.from(previews.keys()).sort();
    const valuesCount = keys.map((k) => previews.get(k)?.length);
    expect(keys).toEqual(['Arrows', 'Emojis', 'Geometries', 'Objects', 'Pins', 'Symbols', 'Weather']);
    expect(valuesCount).toEqual([73, 22, 25, 258, 10, 65, 65]);
    keys
      .flatMap((k) => previews.get(k))
      .forEach((p) => {
        expect(p?.icon.name).toBeDefined();
        expect(p?.icon.category).toBeDefined();
        expect(p?.icon.contentSvg).toBeDefined();
      });
  });
});
