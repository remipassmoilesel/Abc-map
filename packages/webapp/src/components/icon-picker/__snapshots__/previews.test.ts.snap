// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`previews should work 1`] = `
Array [
  "Arrows",
  "Emojis",
  "Geometries",
  "Iso7010",
  "Objects",
  "Pins",
  "Symbols",
  "Weather",
]
`;

exports[`previews should work 2`] = `
Array [
  73,
  22,
  25,
  53,
  258,
  10,
  65,
  65,
]
`;
